components/clarinet-format/src/formatter/mod.rs

```diff
@@ -1981,4 +2000,24 @@ mod tests_formatter {
         let result = format_with_default(src);
         assert_eq!(expected, result);
     }
+
+    #[test]
+    fn format_ast_without_source() {
+        let src = "(define-private (noop) (begin (+ 1 2) (ok true)))";
+        let ast = clarity::vm::ast::parser::v2::parse(src).unwrap();
+        let formatter = ClarityFormatter::new(Settings::default());
+        let expected = formatter.format_file(src);
+        let result = formatter.format_ast(&ast);
+        assert_eq!(result, expected);
+    }
+
+    #[test]
+    fn format_ast_without_source_handle_indentation() {
+        let src = "  (begin (+ 1 2) (ok true))\n";
+        let ast = clarity::vm::ast::parser::v2::parse(src).unwrap();
+        let formatter = ClarityFormatter::new(Settings::default());
+        let expected = formatter.format_file(src);
+        let result = formatter.format_ast(&ast);
+        assert_eq!(result + "\n", expected);
```

Comment:
I'm not exactly sure why `expected` gives an extra `\n` (and not `result`).
Let me know if you have an idea, I'll investigate

Reply:
It's because of the `format_file` rather than `format_section`. `format_file` expects to end with an extra newline